shouldNotLoseDataWhenScalingDown ocasionally fails

A couple of times I have seen this fail. Why?

Re run many times to prove stability.

The issue is due to inconsistent results in the /_count aggregation across shards:
```
[INFO] 2016-01-11 08:17:56,560 org.apache.mesos.elasticsearch.systemtest.ScalingSystemTest lambda$shouldNotLoseDataWhenScalingDown$3 - [{"_shards":{"total":5,"failed":0,"successful":5},"count":1}]
[INFO] 2016-01-11 08:17:56,667 org.apache.mesos.elasticsearch.systemtest.ScalingSystemTest lambda$shouldNotLoseDataWhenScalingDown$3 - [{"_shards":{"total":5,"failed":0,"successful":5},"count":2}]
[INFO] 2016-01-11 08:17:56,789 org.apache.mesos.elasticsearch.systemtest.ScalingSystemTest lambda$shouldNotLoseDataWhenScalingDown$3 - [{"_shards":{"total":5,"failed":0,"successful":5},"count":9}]
[INFO] 2016-01-11 08:17:56,850 org.apache.mesos.elasticsearch.systemtest.ScalingSystemTest lambda$shouldNotLoseDataWhenScalingDown$3 - [{"_shards":{"total":5,"failed":0,"successful":5},"count":2}]
[INFO] 2016-01-11 08:17:56,977 org.apache.mesos.elasticsearch.systemtest.ScalingSystemTest lambda$shouldNotLoseDataWhenScalingDown$3 - [{"_shards":{"total":5,"failed":0,"successful":5},"count":6}]
[INFO] 2016-01-11 08:17:57,096 org.apache.mesos.elasticsearch.systemtest.ScalingSystemTest lambda$shouldNotLoseDataWhenScalingDown$3 - [{"_shards":{"total":5,"failed":0,"successful":5},"count":9}]
[INFO] 2016-01-11 08:17:57,108 org.apache.mesos.elasticsearch.systemtest.ScalingSystemTest lambda$shouldNotLoseDataWhenScalingDown$3 - [{"_shards":{"total":5,"failed":0,"successful":5},"count":9}]
[INFO] 2016-01-11 08:17:57,134 org.apache.mesos.elasticsearch.systemtest.ScalingSystemTest lambda$shouldNotLoseDataWhenScalingDown$3 - [{"_shards":{"total":5,"failed":0,"successful":5},"count":10}]
[DEBUG] 2016-01-11 08:17:57,144 org.apache.mesos.elasticsearch.systemtest.ScalingSystemTest shouldNotLoseDataWhenScalingDown - Correct number of documents: 9
[DEBUG] 2016-01-11 08:17:57,181 org.apache.mesos.elasticsearch.systemtest.ScalingSystemTest shouldNotLoseDataWhenScalingDown - Scaling down to 1
[DEBUG] 2016-01-11 08:17:57,193 org.apache.mesos.elasticsearch.systemtest.ScalingSystemTest shouldNotLoseDataWhenScalingDown - Response 200
[DEBUG] 2016-01-11 08:17:58,228 org.apache.mesos.elasticsearch.systemtest.ScalingSystemTest lambda$shouldNotLoseDataWhenScalingDown$4 - Number of nodes: 1
[DEBUG] 2016-01-11 08:17:58,229 org.apache.mesos.elasticsearch.systemtest.ESTasks getTasks - Fetching tasks on http://172.17.0.7:31100/v1/tasks
[INFO] 2016-01-11 08:17:58,262 org.apache.mesos.elasticsearch.systemtest.ScalingSystemTest shouldNotLoseDataWhenScalingDown - New address: [172.17.0.1:9200]
[DEBUG] 2016-01-11 08:17:58,262 org.apache.mesos.elasticsearch.systemtest.ScalingSystemTest waitForGreen - Wating for green.
[DEBUG] 2016-01-11 08:18:49,952 org.apache.mesos.elasticsearch.systemtest.ESTasks getTasks - Fetching tasks on http://172.17.0.7:31100/v1/tasks
[DEBUG] 2016-01-11 08:18:49,962 org.apache.mesos.elasticsearch.systemtest.ScalingSystemTest lambda$waitForGreen$8 - {"cluster_name":"mesos-ha","status":"green","timed_out":false,"number_of_nodes":1,"number_of_data_nodes":1,"active_primary_shards":5,"active_shards":5,"relocating_shards":0,"initializing_shards":0,"unassigned_shards":0,"delayed_unassigned_shards":0,"number_of_pending_tasks":0,"number_of_in_flight_fetch":0,"task_max_waiting_in_queue_millis":0,"active_shards_percent_as_number":100.0}
[DEBUG] 2016-01-11 08:18:50,973 org.apache.mesos.elasticsearch.systemtest.ScalingSystemTest shouldNotLoseDataWhenScalingDown - New number of documents: 10

java.lang.AssertionError: 
Expected :9
Actual   :10
```

--- a/system-test/src/systemTest/java/org/apache/mesos/elasticsearch/systemtest/ScalingSystemTest.java
+++ b/system-test/src/systemTest/java/org/apache/mesos/elasticsearch/systemtest/ScalingSystemTest.java
@@ -11,6 +11,7 @@ import org.apache.log4j.Logger;
 import org.apache.mesos.elasticsearch.systemtest.base.SchedulerTestBase;
 import org.apache.mesos.elasticsearch.systemtest.containers.DataPusherContainer;
 import org.json.JSONArray;
+import org.junit.Before;
 import org.junit.Test;
 
 import java.util.List;
@@ -20,57 +21,49 @@ import java.util.stream.Collectors;
 import static org.junit.Assert.assertEquals;
 
 /**
- * Tests the scaling capabilities
+ * Tests the scaling capabilities. To run multiple times, uncomment the code below.
  */
+//@RunWith(Parameterized.class)
 public class ScalingSystemTest extends SchedulerTestBase {
     private static final Logger LOGGER = Logger.getLogger(ScalingSystemTest.class);
     public static final int WEBUI_PORT = 31100;
     private static DockerClient dockerClient = DockerClientFactory.build();
+    private String ipAddress;
+    private ESTasks esTasks;
+
+//    @Parameterized.Parameters
+//    public static List<Object[]> data() {
+//        return Arrays.asList(new Object[10][0]);
+//    }
+
+    @Before
+    public void before() {
+        ipAddress = getScheduler().getIpAddress();
+        esTasks = new ESTasks(TEST_CONFIG, ipAddress);
+    }
 
     @Test
     public void shouldScaleDown() throws UnirestException {
-        String ipAddress = getScheduler().getIpAddress();
-        String oldNumNodes = getNumberOfNodes(ipAddress);
-        LOGGER.debug("Number of nodes: " + oldNumNodes);
-
-        String newNumNodes = "2";
-        HttpResponse<JsonNode> response = Unirest.put("http://" + ipAddress + ":" + WEBUI_PORT + "/v1/cluster/elasticsearchNodes").header("Content-Type", "application/json").body(newNumNodes).asJson();
-        assertEquals(200, response.getStatus());
-
-        Awaitility.await().atMost(60, TimeUnit.SECONDS).pollInterval(1, TimeUnit.SECONDS).until(() -> {
-            String numNodes = getNumberOfNodes(ipAddress);
-            LOGGER.debug("Number of nodes: " + numNodes);
-            return numNodes.equals(newNumNodes);
-        });
-        assertEquals(newNumNodes, getNumberOfNodes(ipAddress));
+        scaleNumNodesTo(ipAddress, "3"); // Reset to 3 nodes
+        waitForGreen(esTasks);
+        LOGGER.debug("Number of nodes: " + getNumberOfNodes(ipAddress));
+        scaleNumNodesTo(ipAddress, "2");
+        waitForGreen(esTasks);
     }
 
     @Test
     public void shouldScaleUp() throws UnirestException {
-        shouldScaleDown();
-
-        String ipAddress = getScheduler().getIpAddress();
-        String oldNumNodes = getNumberOfNodes(ipAddress);
-        LOGGER.debug("Number of nodes: " + oldNumNodes);
-
-        String newNumNodes = "3";
-        HttpResponse<JsonNode> response = Unirest.put("http://" + ipAddress + ":" + WEBUI_PORT + "/v1/cluster/elasticsearchNodes").header("Content-Type", "application/json").body(newNumNodes).asJson();
-        assertEquals(200, response.getStatus());
-
-        Awaitility.await().atMost(60, TimeUnit.SECONDS).pollInterval(1, TimeUnit.SECONDS).until(() -> {
-            String numNodes = getNumberOfNodes(ipAddress);
-            LOGGER.debug("Number of nodes: " + numNodes);
-            return numNodes.equals(newNumNodes);
-        });
-        assertEquals(newNumNodes, getNumberOfNodes(ipAddress));
+        scaleNumNodesTo(ipAddress, "2"); // Reset to 2 nodes
+        waitForGreen(esTasks);
+        LOGGER.debug("Number of nodes: " + getNumberOfNodes(ipAddress));
+        scaleNumNodesTo(ipAddress, "3");
+        waitForGreen(esTasks);
     }
 
     @Test
     public void shouldNotLoseDataWhenScalingDown() throws UnirestException {
         // Make sure we have three nodes
-        shouldScaleUp();
-
-        ESTasks esTasks = new ESTasks(TEST_CONFIG, getScheduler().getIpAddress());
+        scaleNumNodesTo(ipAddress, "3");
         waitForGreen(esTasks);
 
         List<String> esAddresses = esTasks.getTasks().stream().map(task -> task.getString("http_address")).collect(Collectors.toList());
@@ -85,12 +78,13 @@ public class ScalingSystemTest extends SchedulerTestBase {
             for (String httpAddress : esAddresses) {
                 try {
                     responseElements = Unirest.get("http://" + httpAddress + "/_count").asJson().getBody().getArray();
+                    LOGGER.info(responseElements);
+                    int count = responseElements.getJSONObject(0).getInt("count");
+                    if (count == 0 || count % 10 != 0) { // There are ten documents in the data pusher. This is for repeated testing, if used.
+                        return false;
+                    }
                 } catch (Exception e) {
                     LOGGER.error("Unirest exception:" + e.getMessage());
-                    return false;
-                }
-                LOGGER.info(responseElements);
-                if (9 > responseElements.getJSONObject(0).getInt("count")) {
                     return false;
                 }
             }
@@ -101,23 +95,13 @@ public class ScalingSystemTest extends SchedulerTestBase {
         LOGGER.debug("Correct number of documents: " + correctNumberOfDocuments);
 
         // Scale down to one node
-        String ipAddress = getScheduler().getIpAddress();
-        String newNumNodes = "1";
-        LOGGER.debug("Scaling down to " + newNumNodes);
-        HttpResponse<JsonNode> response = Unirest.put("http://" + ipAddress + ":" + WEBUI_PORT + "/v1/cluster/elasticsearchNodes").header("Content-Type", "application/json").body(newNumNodes).asJson();
-        LOGGER.debug("Response " + response.getStatus());
-        assertEquals(200, response.getStatus());
-        Awaitility.await().atMost(60, TimeUnit.SECONDS).pollInterval(1, TimeUnit.SECONDS).until(() -> {
-            String numNodes = getNumberOfNodes(ipAddress);
-            LOGGER.debug("Number of nodes: " + numNodes);
-            return numNodes.equals(newNumNodes);
-        });
+        scaleNumNodesTo(ipAddress, "1");
+        waitForGreen(esTasks);
 
         // Refresh ES address list
         List<String> newAddress = esTasks.getTasks().stream().map(task -> task.getString("http_address")).collect(Collectors.toList());
         LOGGER.info("New address: " + newAddress);
         assertEquals(1, newAddress.size());
-        waitForGreen(esTasks);
 
         Awaitility.await().atMost(60, TimeUnit.SECONDS).pollInterval(1, TimeUnit.SECONDS).until(() -> {
             try {
@@ -134,9 +118,21 @@ public class ScalingSystemTest extends SchedulerTestBase {
         assertEquals(correctNumberOfDocuments, newNumberOfDocuments);
     }
 
+    public void scaleNumNodesTo(String ipAddress, String newNumNodes) throws UnirestException {
+        HttpResponse<JsonNode> response = Unirest.put("http://" + ipAddress + ":" + WEBUI_PORT + "/v1/cluster/elasticsearchNodes").header("Content-Type", "application/json").body(newNumNodes).asJson();
+        assertEquals(200, response.getStatus());
+
+        Awaitility.await().atMost(60, TimeUnit.SECONDS).pollInterval(1, TimeUnit.SECONDS).until(() -> {
+            String numNodes = getNumberOfNodes(ipAddress);
+            LOGGER.debug("Number of nodes: " + numNodes);
+            return numNodes.equals(newNumNodes);
+        });
+        assertEquals(newNumNodes, getNumberOfNodes(ipAddress));
+    }
+
     public void waitForGreen(ESTasks esTasks) {
         LOGGER.debug("Wating for green.");
-        Awaitility.await().atMost(5, TimeUnit.MINUTES).pollDelay(2, TimeUnit.SECONDS).until(() -> {
+        Awaitility.await().atMost(5, TimeUnit.MINUTES).pollInterval(1, TimeUnit.SECONDS).until(() -> { // This can take some time, somtimes.
             try {
                 List<String> esAddresses = esTasks.getTasks().stream().map(task -> task.getString("http_address")).collect(Collectors.toList());
                 // This may throw a JSONException if we call before the JSON has been generated. Hence, catch exception.

--- a/system-test/src/systemTest/java/org/apache/mesos/elasticsearch/systemtest/ScalingSystemTest.java
+++ b/system-test/src/systemTest/java/org/apache/mesos/elasticsearch/systemtest/ScalingSystemTest.java
@@ -44,26 +44,26 @@ public class ScalingSystemTest extends SchedulerTestBase {
 
     @Test
     public void shouldScaleDown() throws UnirestException {
-        scaleNumNodesTo(ipAddress, "3"); // Reset to 3 nodes
+        scaleNumNodesTo(ipAddress, 3); // Reset to 3 nodes
         waitForGreen(esTasks);
         LOGGER.debug("Number of nodes: " + getNumberOfNodes(ipAddress));
-        scaleNumNodesTo(ipAddress, "2");
+        scaleNumNodesTo(ipAddress, 2);
         waitForGreen(esTasks);
     }
 
     @Test
     public void shouldScaleUp() throws UnirestException {
-        scaleNumNodesTo(ipAddress, "2"); // Reset to 2 nodes
+        scaleNumNodesTo(ipAddress, 2); // Reset to 2 nodes
         waitForGreen(esTasks);
         LOGGER.debug("Number of nodes: " + getNumberOfNodes(ipAddress));
-        scaleNumNodesTo(ipAddress, "3");
+        scaleNumNodesTo(ipAddress, 3);
         waitForGreen(esTasks);
     }
 
     @Test
     public void shouldNotLoseDataWhenScalingDown() throws UnirestException {
         // Make sure we have three nodes
-        scaleNumNodesTo(ipAddress, "3");
+        scaleNumNodesTo(ipAddress, 3);
         waitForGreen(esTasks);
 
         List<String> esAddresses = esTasks.getTasks().stream().map(task -> task.getString("http_address")).collect(Collectors.toList());
@@ -95,7 +95,7 @@ public class ScalingSystemTest extends SchedulerTestBase {
         LOGGER.debug("Correct number of documents: " + correctNumberOfDocuments);
 
         // Scale down to one node
-        scaleNumNodesTo(ipAddress, "1");
+        scaleNumNodesTo(ipAddress, 1);
         waitForGreen(esTasks);
 
         // Refresh ES address list
@@ -118,16 +118,16 @@ public class ScalingSystemTest extends SchedulerTestBase {
         assertEquals(correctNumberOfDocuments, newNumberOfDocuments);
     }
 
-    public void scaleNumNodesTo(String ipAddress, String newNumNodes) throws UnirestException {
-        HttpResponse<JsonNode> response = Unirest.put("http://" + ipAddress + ":" + WEBUI_PORT + "/v1/cluster/elasticsearchNodes").header("Content-Type", "application/json").body(newNumNodes).asJson();
+    public void scaleNumNodesTo(String ipAddress, Integer newNumNodes) throws UnirestException {
+        HttpResponse<JsonNode> response = Unirest.put("http://" + ipAddress + ":" + WEBUI_PORT + "/v1/cluster/elasticsearchNodes").header("Content-Type", "application/json").body(newNumNodes.toString()).asJson();
         assertEquals(200, response.getStatus());
 
         Awaitility.await().atMost(60, TimeUnit.SECONDS).pollInterval(1, TimeUnit.SECONDS).until(() -> {
             String numNodes = getNumberOfNodes(ipAddress);
             LOGGER.debug("Number of nodes: " + numNodes);
-            return numNodes.equals(newNumNodes);
+            return numNodes.equals(newNumNodes.toString());
         });
-        assertEquals(newNumNodes, getNumberOfNodes(ipAddress));
+        assertEquals(newNumNodes.toString(), getNumberOfNodes(ipAddress));
     }
 
     public void waitForGreen(ESTasks esTasks) {

